### PR TITLE
Update subscribe and assign method to follow same processing logic as apache kafka

### PR DESCRIPTION
**Pull request by haijun2022**
This PR fixed the bug described in issue https://github.com/strimzi/strimzi-kafka-bridge/issues/756

it update subscribe and assign method to follow same processing logic as apache kafka.

### CODE REVIEW

File: src/main/java/io/strimzi/kafka/bridge/KafkaBridgeConsumer.java [hunk 1]
```diff
@@ -93,8 +93,12 @@ public void create(Properties config, String groupId) {
      * @param topicSubscriptions topics to subscribe to
      */
     public void subscribe(List<SinkTopicSubscription> topicSubscriptions) {
+        if (topicSubscriptions == null) {
+            throw new IllegalArgumentException("Topic subscriptions cannot be null");
+        }
+
         if (topicSubscriptions.isEmpty()) {
-            throw new IllegalArgumentException("At least one topic to subscribe has to be specified!");
+            this.unsubscribe();
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/KafkaBridgeConsumer.java:101]: so after unsubscribing you are actually not returning and the subscribe operation will run anyway. It should not cause a problem just because the Kafka subscribe with empty topic list will behave as an unsubscribe.
For this reason I would say that this addition about calling unsubscribe is not needed, we should just allow an empty topics list and then allowing the code to run the subscribe.

File: src/main/java/io/strimzi/kafka/bridge/KafkaBridgeConsumer.java [hunk 2]
```diff
@@ -151,6 +155,10 @@ public void assign(List<SinkTopicSubscription> topicSubscriptions) {
             topicPartitions.add(new TopicPartition(topicSubscription.getTopic(), topicSubscription.getPartition()));
         }
 
+        if (topicPartitions.isEmpty()) {
+            this.unsubscribe();
+        }
```
Comment (ppatierno) [inline src/main/java/io/strimzi/kafka/bridge/KafkaBridgeConsumer.java:160]: same as described before. Run the unsubscribe here, not returning and then the code runs the assign.
Just running the assign with an empty topics list should work.

**Review by ppatierno — CHANGES_REQUESTED** [review]

File: src/main/java/io/strimzi/kafka/bridge/KafkaBridgeConsumer.java (hunk 2 above)
Comment (haijun2022) [reply src/main/java/io/strimzi/kafka/bridge/KafkaBridgeConsumer.java:160]: Thanks for the review, exactly, will update the code!

**Comment by haijun2022** [comment]
@ppatierno @scholzj I have updated the code and added test code based on your suggestions. Could you reivew it again and let me know if there is any suggestions. thank you!

File: src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java [hunk 3]
```diff
@@ -160,6 +161,62 @@ void subscriptionConsumerDoesNotExistBecauseNotCreated(VertxTestContext context)
...
+        JsonObject topicsRoot = new JsonObject();
+        topicsRoot.put("topics", topics);
+
+        // Create consumer and subscribe to the topic
+        JsonObject consumerJson = new JsonObject();
+        consumerJson.put("name", name);
+        consumerJson.put("format", "json");
+        consumerService()
+                .createConsumer(context, groupId, consumerJson)
+                .subscribeConsumer(context, groupId, name, topicsRoot);
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:181]: after subscribing you should verify that it was ok by listing the subscription, so that after unsubscribe you can do the same and checking it's empty.
Anyway, for how the Kafka client works, after the subscribe operation you need to execute some poll/consume in order to allow the consumer to be really subscribed (joining the consumer group). The subscribe operation itself just fill a list inside the Kafka consumer nothing more.
See the `listConsumerSubscriptions` test which does something like that after subscribe and before listing the subscriptions.

File: src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java [hunk 4]
```diff
@@ -355,6 +412,60 @@ void assignAfterSubscriptionTest(VertxTestContext context) throws InterruptedExc
...
+        String consumerName = "my-kafka-consumer-assign-empty";
+
+        JsonObject consumerJson = new JsonObject();
+        consumerJson.put("name", consumerName);
+        consumerJson.put("format", "json");
+
+        JsonObject partitionsRoot = new JsonObject();
+        JsonArray partitions = new JsonArray();
+
+        partitionsRoot.put("partitions", partitions);
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:431]: compared to the previous test, there is no assignment to a partition and then assignment with empty list to validate the feature.

**Review by ppatierno — CHANGES_REQUESTED** [review]

**Comment by ppatierno** [comment]
@haijun2022 I see you are on the right way! I noticed you added the check I suggested on the subscription test only but it should be on the assign test as well.

**Comment by haijun2022** [comment]
> @haijun2022 I see you are on the right way! I noticed you added the check I suggested on the subscription test only but it should be on the assign test as well.

Thank you for the review, I tried but did not find a way to go to that logic branch as you suggested so I removed that test code! Will double check, meanwhile please let me know if that is doable!

**Comment by ppatierno** [comment]
> Thank you for the review, I tried but did not find a way to go to that logic branch as you suggested so I removed that test code! Will double check, meanwhile please let me know if that is doable!

I am not sure what you mean by this :-(
You added the logic to unsubscribe when someone use `assign` but with empty partitions.
You added a test for it. But it was missing the suggestions I made to you on the other test (the one about unsubscribe via a `subscribe` with empty topics collection).

**Comment by ppatierno** [comment]
@haijun2022 do you have any idea when this PR will be ready? We are in the process to start a 0.25.0 release and having your contribution included would be great :-)

**Comment by haijun2022** [comment]
Hi @ppatierno, I am working on it now and will update by the end of today, thank you!

File: src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java (hunk 4 above)
Comment (haijun2022) [reply src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:470]: just added the section, thank you!

**Comment by haijun2022** [comment]
> Hi @ppatierno, I am working on it now and will update by the end of today, thank you!

Hi @ppatierno , I just updated the code and please help review on it.  My IDE added some whitespaces which is not my intension, please ignore them during your review. Sorry for that!

**Comment by ppatierno** [comment]
@haijun2022 it would be great you remove those whitespaces. They are noise in the PR and we don't want to merge the PR with them, thanks!

**Comment by haijun2022** [comment]
> 

Yes, let me know if others are good now? I will update them later

File: src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java [hunk 5]
```diff
@@ -160,6 +161,92 @@ void subscriptionConsumerDoesNotExistBecauseNotCreated(VertxTestContext context)
...
+                .as(BodyCodec.jsonArray())
+                .send(ar -> {
+                    if (ar.succeeded()) {
+                        LOGGER.info("Request result: {}", ar.result().body());
+                        consume.complete(true);
+                    }
+                });
+        consume.get(TEST_TIMEOUT, TimeUnit.SECONDS);
+
+        CompletableFuture<Boolean> subscribe = new CompletableFuture<>();
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:197]: can you create this later when it is really used just before the subscription to an empty topics list? Maybe even calling `subscribeEmpty` would be better. Thanks!

File: src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java [hunk 6]
```diff
@@ -160,6 +161,92 @@ void subscriptionConsumerDoesNotExistBecauseNotCreated(VertxTestContext context)
...
+                        HttpResponse<JsonObject> response = ar.result();
+                        assertThat(response.statusCode(), is(HttpResponseStatus.NO_CONTENT.code()));
+                        assertThat(response.body(), is(nullValue()));
+                    });
+                    subscribe.complete(true);
+                });
+
+        subscribe.get(TEST_TIMEOUT, TimeUnit.SECONDS);
+        // Validate the existing consumer list
+        CompletableFuture<Boolean> listSubscriptionsAfter = new CompletableFuture<>();
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:231]: you can reuse the `listSubscriptions` variable as before just creating a `new CompletableFuture<>();`

File: src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java [hunk 7]
```diff
@@ -355,6 +442,105 @@ void assignAfterSubscriptionTest(VertxTestContext context) throws InterruptedExc
...
+                        HttpResponse<JsonObject> response = ar.result();
+                        assertThat(response.statusCode(), is(HttpResponseStatus.NO_CONTENT.code()));
+                    });
+                    assignEmptyCF.complete(true);
+                });
+
+        assignEmptyCF.get(TEST_TIMEOUT, TimeUnit.SECONDS);
+
+        // Validate the existing consumer list
+        CompletableFuture<Boolean> listSubscriptionsAfter = new CompletableFuture<>();
```
Comment (ppatierno) [inline src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:524]: same as before you should be able to use just one `listSubscriptions` completable future for before and after, just creating a new instance assigning to the same variable.

**Comment by ppatierno** [comment]
@haijun2022 I think these are last comments and we are good to go.

File: src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java (hunk 7 above)
Comment (haijun2022) [reply src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:524]: Hi @ppatierno , that does not work here. it shows "Local variable listSubscriptionsBefore defined in an enclosing scope must be final or effectively finalJava(536871575)"

File: src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java (hunk 5 above)
Comment (haijun2022) [reply src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java]: updated, thanks!

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks a lot for the contribution!

**Comment by haijun2022** [comment]
> LGTM. Thanks a lot for the contribution!

Thanks a lot for your reivew and suggestions. @ppatierno

**Review by scholzj — APPROVED** [review]
